data_model/src/lib.rs

```diff
@@ -864,6 +866,113 @@ mod model {
         /// in the next request to continue fetching results of the original query
         pub cursor: crate::query::cursor::ForwardCursor,
     }
+
+    /// String containing serialized valid JSON.
+    ///
+    /// This string is guaranteed to be parsed as JSON.
+    #[derive(Display, Default, Debug, Clone, Eq, Encode, Decode, TypeId, Ord, PartialOrd)]
+    #[ffi_type(unsafe {robust})]
+    #[repr(transparent)]
+    #[display(fmt = "{}", "0")]
+    pub struct JsonString(pub(super) String);
+}
+
+/// A helper trait for polymorphism, implemented for various types
+pub trait IntoJsonString {
+    /// Converts self into [`JsonString`]
+    fn into_json_string(self) -> JsonString;
+}
+
+impl IntoJsonString for JsonString {
+    fn into_json_string(self) -> JsonString {
+        self
+    }
+}
+
+impl IntoJsonString for &serde_json::Value {
+    fn into_json_string(self) -> JsonString {
+        JsonString::from(self)
+    }
+}
+
+impl IntoJsonString for serde_json::Value {
+    fn into_json_string(self) -> JsonString {
+        (&self).into_json_string()
+    }
+}
+
+impl JsonString {
+    /// Deserialize JSON into something
+    /// # Errors
+    /// See [`serde_json::from_str`].
+    pub fn deserialize<'a, T>(&'a self) -> serde_json::Result<T>
+    where
+        T: Deserialize<'a>,
+    {
+        serde_json::from_str(&self.0)
+    }
+
+    /// Serializes a value into [`JsonString`]
+    /// # Errors
+    /// See [`serde_json::to_string`].
+    pub fn serialize<T: serde::Serialize>(value: &T) -> serde_json::Result<Self> {
+        let serialized = serde_json::to_string(value)?;
+        // the string was obtained from serde_json serialization,
+        // so it should be a valid JSON string
+        Ok(Self(serialized))
+    }
+
+    /// Create without checking whether the input is a valid JSON string.
+    ///
+    /// The caller must guarantee that the value is valid.
+    pub fn from_json_string_unchecked(value: String) -> Self {
```

Comment:
if you want to have this method, mark it as `transparent_api` or sth

Reply:
Why?

Reply:
I just didn't want to have it in the public API. Maybe this method can take `JsonString` and `JsonString` to have unchecked constructor?

Reply:
because we'll use JsonString for different things

Reply:
> Maybe this method can take `JsonString` and `JsonString` to have unchecked constructor?

I don't get it... This is already an unchecked constructor of `JsonString` - `JsonString::from_json_string_unchecked`.

Reply:
my bad. Then:
```suggestion
    pub fn from_string_unchecked(value: String) -> Self {
```